New app: luci-app-sshtunnel for SSH tunnels

stokito opened the pull request:
@nunojpg please have a look
CC @stweil @kvuorine @kzyapkov you guys also contributed to the sshtunnel.

![Tunnels Tab](https://github.com/openwrt/luci/assets/415502/674a8c4a-8257-444a-b2da-6e690541a203)


Notes:
Since in future the Dropbear's `dbclient` will be supported I moved all `openssh-client` only options to advanced tab. And I didn't added some options like Compression Level simply because   their defaults is fine for 99.999% of users.

I added Keys section that allows to generate a key and copy a pub key.
A user may not have `ssh-keygen` if removed the `openssh-client` then Dropbear's keygen will be used instead. It will run a slightly different command but in future they may be same https://github.com/mkj/dropbear/issues/236

stokito commented:
The user and port fields are optional as I described here https://github.com/openwrt/packages/pull/21263/commits/cc18840ddd9e5c060d9b4333c7d0ec6b00a004e3

The remote address field is also can be empty. Added a new Enabled option.


I also added Known Hosts table. Not sure if it makes sense but at least it will give more info for a user to understand what's happens. This is useful for security perspective.

jow- commented:
It looks a little bit cluttered. Would it make sense to move the various tunnel types + key management into separate tabs?

stokito commented:
@jow- that's a good point and I thought about it but decided to leave it as is for few reasons:
1. I'm pretty sure that 99% of users will just setup one R tunnel and one server and have one key. So the page won't be so overloaded with info.
2. IMHO this will make it easier to write and instruction: click here, then here instead of click here and switch tab and click here. I think it would be more user friendly especially given that they all look same and simple. But if we'll have some additional options then we can split to tabs.

The page should be opened one time and this is fine if it's ugly a little bit.
I tried to minimize UI elements so we don't have a Delete key button or selection of a key type and it's always assumed to be ed25519 (it's already widely supported)

stokito commented:
The bigger UI problem is that after adding a server a user should save settings and only then it can see the new server in tunnels dropdowns. E.g. we need to dynamically add them.
I tried to find how to fix it in existing Luci sources but didn't found anything.

Anyway, you a free to change the UI as you think it would be great. I don't have a time today to change anything.

systemcrash commented:
+1 for tabs. tabs highlight the context. Single page loads are nice, so I understand your motivation. Otherwise, tunnels is a nice addition.

stokito commented:
@jow- if I'll split to tabs, will you accept the PR?

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+
+
+		s = m.section(form.GridSection, "tunnelR", _("Remote Tunnels"),
+			_("When the connection initiated at RemoteAddress:RemotePort and then forwarded to LocalAddress:LocalPort")
```
Your wording is unclear. 
Try:

`_("Forward a port or ports on remote systems to the local system.")`

systemcrash replied:
I gave proposals for those already.

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+		o = _addServerOption(s);
+
+		o = s.option(form.Value, "remoteaddress", _("Remote address"),
+			_("* means to accept a connection from any interface on the OpenWrt side")
```
`_("<em>*</em> means any address.")`

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+
+
+		s = m.section(form.GridSection, "tunnelL", _("Local Tunnels"),
+			_("When the connection initiated at LocalAddress:LocalPort and then forwarded to RemoteAddress:RemotePort")
```
Your wording is unclear. 
Try:

`_("Forward a port or ports on the local system to remote systems.")`

stokito replied:
"Forward a port on the local host to a service on the remote host."

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+		o = _addServerOption(s);
+
+		o = s.option(form.Value, "localaddress", _("Local address"),
+			_("* means to accept a connection from any interface on the OpenWrt side")
```
`_("<em>*</em> means any local address.")`

stokito replied:
I used "Bind IP address or hostname. <em>*</em> means to listen all interfaces <b>including public</b>." The addition about public interfaces is important for security.

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+
+
+		s = m.section(form.GridSection, "tunnelD", _("Dynamic Tunnels"),
+			_("When the connection initiated with the SOCKS protocol to the LocalAddress:LocalPort and then forwarded over the remote host")
```
What?

`_("Forward local ports initiated via SOCKS protocol to a remote host.")`

stokito replied:
What about just "SOCKS proxy via remote host."?

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+		o = _addServerOption(s);
+
+		o = s.option(form.Value, "localaddress", _("Local address"),
+			_("* means to accept a connection from any interface on the OpenWrt side")
```
`_("<em>*</em> means any local address.")`

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+
+
+		s = m.section(form.GridSection, "tunnelW", _("VPN Tunnels"),
+			_("Creates TUN/TAP devices on client and server to establish a VPN tunnel between them")
```
`_("Configure TUN/TAP devices for VPN tunnels.")`

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+});
+
+function _addServerOption(s) {
+	var o = s.option(form.ListValue, "server", _("Server"), _("Type a server name from servers list above"));
```
`var o = s.option(form.ListValue, "server", _("Server"), _("Enter a server name from the server list above."));`

stokito replied:
fixed

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+	return E('div', {'class': 'cbi-section cbi-tblsection'}, [
+		E('h3', _('Keys')),
+		E('div', {'class': 'cbi-section-descr'},
+			_("You need to add the pub key to /root/.ssh/authorized_keys or to /etc/dropbear/authorized_keys.") + " " +
```
`_("Add the pub key to <em>/root/.ssh/authorized_keys</em> or <em>/etc/dropbear/authorized_keys</em>.") + " " +`

stokito replied:
fixed

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+function _handleKeyGen(event) {
+	var keyName = document.querySelector('input[name="keyName"]').value;
+	if (!keyName) {
+		alert(_("Please specify a key name"));
```
`return _("Specify a key name.");`

DO NOT USE `alert()`.

stokito replied:
This is an onclick event so we can't return anything. What's wrong with the alert()?

stokito replied:
changed to uiNotification

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+		return;
+	}
+	if (allSshKeys[keyName]) {
+		alert(_("A key with the name already exists"));
```
`return _("A key with that name already exists.");`

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+		return;
+	}
+	if (/[^a-zA-Z0-9_.]/.test(keyName)) {
+		alert(_("Bad name"));
```
return... not alert

stokito replied:
I removed the check but put it into the input pattern

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+	}
+	return fs.exec(command, commandArgs).then(function (res) {
+		if (res.code === 0) {
+			alert("A new key was generated. Please refresh the page in a browser to see it.");
```
Just `return;` - handle failures, not success.

stokito replied:
Here I need to say "Please refresh the page in a browser to see it" otherwise a user won't see it

stokito replied:
I decided just reload the page instead.

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+			throw new Error(res.stdout + " " + res.stderr);
+		}
+	}).catch(function (err) {
+		alert(_("Unable to generate a key: %s").format(err.message));
```
`return ...`

stokito replied:
That's an error, a user must know about it. But maybe I can use notification instead

stokito replied:
I changed to `ui.addNotification()`

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+	return E('div', {'class': 'cbi-section cbi-tblsection'}, [
+		E('h3', _('Known hosts ')),
+		E('div', {'class': 'cbi-section-descr'},
+			_("Keys of SSH servers from /root/.ssh/known_hosts.")
```
`_("SSH server keys from <em>/root/.ssh/known_hosts</em>.")`

stokito replied:
applied, ty

systemcrash reviewed (commented):
Your `.po` files will need updating to sync with the recommended changes.

stokito commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+		o.default = "root";
+
+		o = s.taboption("general", form.ListValue, "IdentityFile", _("Key file"),
+			_("Private key file with authentication identity. The default is /root/.ssh/id_ed25519, id_rsa, id_ecdsa or id_dropbear.")
```
I decided to remove the default mention

stokito commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+
+		m = new form.Map("sshtunnel", _("SSH Tunnels"),
+			_("This configures <a href='https://openwrt.org/docs/guide-user/services/ssh/sshtunnel'>SSH Tunnels</a>. " +
+				"Password auth is not possible so only Public Key auth must be used.")
```
I feel that so is better

systemcrash replied:
Sounds weird.

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+		var sshKeys = _splitSshKeys(sshFiles.splice(3));
+		if (sshKeys.length === 0) {
+			ui.addNotification(null, E("p", [
+				_("No any SSH key, please generate a new one")
```
I make the recommendations as a native English speaker. The grammar in your PR make the writer sound...insane. Not least difficult to understand for those who don't have English as a first language where they sometimes have to guess at meaning. Поищи с помощью Гугла "No any". Ты пытаешься использовать грамматику которой нет.

stokito replied:
"No SSH keys found, generate a new one" would be fine?

stokito commented:
I splitted to four tabs:
- SSH Keys
- Servers
- Tunnels
- Known Hosts

Also I fixed and simplified translations as @systemcrash proposed.
Thank you!

stokito commented:
> The bigger UI problem is that after adding a server a user should save settings and only then it can see the new server in tunnels dropdowns. E.g. we need to dynamically add them.

This is not an issue anymore because when opening the Tunnels tab it will load even unsaved servers.

systemcrash commented:
> I splitted to four tabs:
> 
>     * SSH Keys
> 
>     * Servers
> 
>     * Tunnels
> 
>     * Known Hosts
> 

Could you update the header screenshot of your fresh results, please? Expand your window and trim away other desktop to include only the GUI. Also include a few of how various interaction processes look, like adding a tunnel.

systemcrash commented:
I also think that it might be a good idea to link to the (sys->admin->ssh) SSH key page. It's pretty much guaranteed to be on every openwrt install.

stokito commented:
![Tunnels Tab](https://github.com/openwrt/luci/assets/415502/674a8c4a-8257-444a-b2da-6e690541a203)

systemcrash commented:
Replace the picture in the first post - that's the recommended way of doing things - then people see the PR as it currently is, in the first post at the top of the page and don't need to scroll through pages of comments to see the latest. 

Your "When the connection..." description strings are still unintelligible. Use my recommendations there.

stokito commented:
@systemcrash I updated. If you can reword the description of rtunnel then please leave as comment. Anyway it should be fine to merge it even in a current form.

systemcrash commented:
You cannot link to an IP and hope it is the same for all users. Just link to the path. Remove the IP portion.

stokito commented:
It will be almost like the same. And the idea is to copy the link and send to someone with instruction how to add the pub key

systemcrash commented:
No. Look at how other luci elements compose links to on device resources. No copy paste required.

stokito commented:
I removed the IP. Now it looks like:
<a href='/cgi-bin/luci/admin/system/admin/sshkeys'>System / Administration
/ SSH-Keys</a>

On Wed, Aug 2, 2023 at 7:45 PM Paul Dee ***@***.***> wrote:

> No. Look at how other luci elements compose links to on device resources.
> No copy paste required.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/luci/pull/6424#issuecomment-1662572185>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADFODWO67HASNR27T2EVGDXTJ72BANCNFSM6AAAAAAY7IX3AA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Sergey Ponomarev <https://linkedin.com/in/stokito>,
stokito.com

stokito commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+		o.modalonly = true;
+
+		o = s.taboption("advanced", form.ListValue, "VerifyHostKeyDNS", _("Verify Host Key DNS"),
+			_("VerifyHostKeyDNS Verify the remote key using DNS and SSHFP resource records.")
```
I removed the VerifyHostKeyDNS option from GUI. It's highly unlikely that it will be used and it's anyway disabled by default. If anyone wants to use it they still can do this with UCI.

stokito commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+		o.modalonly = true;
+
+		o = s.taboption("advanced", form.Value, "retrydelay", _("Retry delay"),
+			_("Delay after a connection failure before trying to reconnect..")
```
That was already fixed

stokito commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+		o.modalonly = true;
+
+		o = s.taboption("advanced", form.Value, "ServerAliveCountMax", _("Server Alive Count Max"),
+			_("ServerAliveCountMax The number of server alive messages which may be sent before ssh disconnect from the server.")
```
fixed

stokito commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+		o.modalonly = true;
+
+		o = s.taboption("advanced", form.ListValue, "StrictHostKeyChecking", _("Strict host key checking"),
+			_("StrictHostKeyChecking Refuse to connect to hosts if it's key has changed."));
```
Changed to "Refuse to connect to hosts whose host key has changed" as it written in `man ssh_config`

stokito commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel.js:
```diff
@@ -0,0 +1,403 @@
+		o.modalonly = true;
+
+		o = s.taboption("advanced", form.Value, "ServerAliveInterval", _("Server Alive Count Max"),
+			_("ServerAliveInterval Timeout interval in seconds for keep alive.")
```
fixed

stokito commented:
I fixed code style and used ' instead of ".
I think that the PR is ready to be merged

systemcrash commented:
Thoughts? @jow- @hnyman

nunojpg commented:
@jow- I merged the package changes. I think its ok to merge this also.

systemcrash commented:
@hnyman review requested plz

systemcrash commented:
@stokito can you update the screenshot of how this currently appears in luci?

@jow- approved?

stokito commented:
The screenshot in the issue is fresh e.g. four tabs. I can prepare a test package so you can easily install it on your router.

stokito commented:
FYI: I updated the Wiki page abiut the sshtunnel
https://openwrt.org/docs/guide-user/services/ssh/sshtunnel

Hope the app will be merged soon because it's really helpful to bypass NAT

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_hosts.js:
```diff
@@ -0,0 +1,63 @@
+		var m, s, o;
+
+		m = new form.Map('sshtunnel', _('SSH Tunnels'),
+			_('This configures <a href="https://openwrt.org/docs/guide-user/services/ssh/sshtunnel">SSH Tunnels</a>.')
```
Please separate this link out to a `.format()`
This is how I did it for another comment:
`			_('Discard upstream responses containing <a href="%s">RFC1918</a> addresses.').format('https://datatracker.ietf.org/doc/html/rfc1918'));
`

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_keys.js:
```diff
@@ -0,0 +1,128 @@
+		var m, s, o;
+
+		m = new form.Map('sshtunnel', _('SSH Tunnels'),
+			_('This configures <a href="https://openwrt.org/docs/guide-user/services/ssh/sshtunnel">SSH Tunnels</a>.')
```
Please separate this link out to a `.format()`
This is how I did it for another comment:
`			_('Discard upstream responses containing <a href="%s">RFC1918</a> addresses.').format('https://datatracker.ietf.org/doc/html/rfc1918'));
`

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_keys.js:
```diff
@@ -0,0 +1,128 @@
+		E('h3', _('SSH Keys')),
+		E('div', {'class': 'cbi-section-descr'},
+			_('Add the pub key to <em>/root/.ssh/authorized_keys</em> or <em>/etc/dropbear/authorized_keys</em>.') + ' ' +
+			_('In LuCI you can do that with <a href="/cgi-bin/luci/admin/system/admin/sshkeys">System / Administration / SSH-Keys</a>')
```
Please separate this link out to a `.format()`
This is how I did it for another comment:
`			_('Discard upstream responses containing <a href="%s">RFC1918</a> addresses.').format('https://datatracker.ietf.org/doc/html/rfc1918'));
`

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_keys.js:
```diff
@@ -0,0 +1,128 @@
+	return E('div', {'class': 'cbi-section cbi-tblsection'}, [
+		E('h3', _('SSH Keys')),
+		E('div', {'class': 'cbi-section-descr'},
+			_('Add the pub key to <em>/root/.ssh/authorized_keys</em> or <em>/etc/dropbear/authorized_keys</em>.') + ' ' +
```
Sorry, I got this wrong earlier, can you change `em` to `code` please?

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_servers.js:
```diff
@@ -0,0 +1,144 @@
+		var m, s, o;
+
+		m = new form.Map('sshtunnel', _('SSH Tunnels'),
+			_('This configures <a href="https://openwrt.org/docs/guide-user/services/ssh/sshtunnel">SSH Tunnels</a>.')
```
Please separate this link out to a `.format()`
This is how I did it for another comment:
`			_('Discard upstream responses containing <a href="%s">RFC1918</a> addresses.').format('https://datatracker.ietf.org/doc/html/rfc1918'));
`

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_servers.js:
```diff
@@ -0,0 +1,144 @@
+
+		o = s.taboption('general', form.ListValue, 'IdentityFile', _('Key file'),
+			_('Private key file with authentication identity. ' +
+				'See <em>ssh_config IdentityFile</em>')
```
please also change `em` to `code` here

systemcrash replied:
Did this also get updated? Ensure all files get the `<em>` to `<code>` treatment

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_servers.js:
```diff
@@ -0,0 +1,144 @@
+		o.optional = true;
+
+
+		o = s.taboption('advanced', form.ListValue, 'LogLevel', _('Log level'), 'See <em>ssh_config LogLevel</em>');
```
and here

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_servers.js:
```diff
@@ -0,0 +1,144 @@
+
+		o = s.taboption('advanced', form.ListValue, 'Compression', _('Use compression'),
+			_('Compression may be useful on slow connections. ' +
+				'See <em>ssh_config Compression</em>')
```
and here...

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_servers.js:
```diff
@@ -0,0 +1,144 @@
+
+		o = s.taboption('advanced', form.Value, 'ServerAliveCountMax', _('Server alive count max'),
+			_('The number of server alive messages which may be sent before SSH disconnects from the server. ' +
+				'See <em>ssh_config ServerAliveCountMax</em>')
```
and here...

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_servers.js:
```diff
@@ -0,0 +1,144 @@
+
+		o = s.taboption('advanced', form.Value, 'ServerAliveInterval', _('Server alive interval'),
+			_('Keep-alive interval (seconds). ' +
+				'See <em>ssh_config ServerAliveInterval</em>')
```
and here...

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_servers.js:
```diff
@@ -0,0 +1,144 @@
+		o = s.taboption('advanced', form.ListValue, 'CheckHostIP', _('Check host IP'),
+			_('Check the host IP address in the <em>known_hosts</em> file. ' +
+				'This allows ssh to detect whether a host key changed due to DNS spoofing. ' +
+				'See <em>ssh_config CheckHostIP</em>')
```
and here

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_servers.js:
```diff
@@ -0,0 +1,144 @@
+
+		o = s.taboption('advanced', form.ListValue, 'StrictHostKeyChecking', _('Strict host key checking'),
+			_('Refuse to connect to hosts whose host key has changed. ' +
+				'See <em>ssh_config StrictHostKeyChecking</em>'));
```
and here

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_tunnels.js:
```diff
@@ -0,0 +1,163 @@
+		var m, s, o;
+
+		m = new form.Map('sshtunnel', _('SSH Tunnels'),
+			_('This configures <a href="https://openwrt.org/docs/guide-user/services/ssh/sshtunnel">SSH Tunnels</a>.')
```
Please separate this link out to a `.format()`
This is how I did it for another comment:
`			_('Discard upstream responses containing <a href="%s">RFC1918</a> addresses.').format('https://datatracker.ietf.org/doc/html/rfc1918'));
`

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_tunnels.js:
```diff
@@ -0,0 +1,163 @@
+
+		o = s.option(form.Value, 'remoteaddress', _('Remote address'),
+			_('Bind IP address e.g. <em>192.168.1.1</em> or hostname e.g. <em>localhost</em>.') + '<br/>' +
+			_('<em>*</em> means to listen all interfaces <b>including public</b>.')
```
and here

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_tunnels.js:
```diff
@@ -0,0 +1,163 @@
+		o.rmempty = false;
+
+		o = s.option(form.Value, 'localaddress', _('Local address'),
+			_('Bind IP address e.g. <em>192.168.1.1</em> or hostname e.g. <em>localhost</em>.')
```
and here

systemcrash commented on applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_tunnels.js:
```diff
@@ -0,0 +1,163 @@
+		o.rmempty = false;
+
+		o = s.option(form.Value, 'remoteaddress', _('Remote address'),
+			_('Bind IP address e.g. <em>192.168.1.1</em> or hostname e.g. <em>localhost</em>.')
```
and here

systemcrash commented:
Fix those changes and LGTM

jow- commented:
This overlaps in functionality with https://github.com/openwrt/luci/tree/master/protocols/luci-proto-pppossh

stokito commented:
yes, and for the full VPN feature I would recommend the pppossh. But if you just need a port forwading or socks proxy the the sshtunnel will be more productive.
On the wiki I mentioned it https://openwrt.org/docs/guide-user/services/ssh/sshtunnel

stokito commented:
@systemcrash The only `<em>` left in descriptions like <em>ssh_config StrictHostKeyChecking</em>. This not kind of code but a name of a section in the man reference.

systemcrash commented:
> @systemcrash The only `\<em\>` left in descriptions like _ssh_config StrictHostKeyChecking_. This not kind of code but a name of a section in the man reference.

Sure, and that won't change. So that can also go in a format block.

stokito commented:
Thank you all

imayoda commented:
Sry for bothering, is there a plan to backport this very nice app to older openwrt versions ? let's say 19.07 :blush: 
thanks in advance

imayoda commented:
thanks a lot for the reply and kind suggestion. 
Basically I belive 19.07 is the last one usable with some extra packets on 8MB routers, newer versions are big even with only luci installed :)
Have a great day
